ledger/txtail.go

```diff
@@ -43,6 +44,12 @@ func (t *txTail) loadFromDisk(l ledgerForTracker) error {
 	}
 	proto := config.Consensus[hdr.CurrentProtocol]
 
+	t.lastValid = make(map[basics.Round]map[transactions.Txid]struct{})
+	for i := uint64(0); i <= proto.MaxTxnLife+1; i++ {
```

Comment:
Why `+1` here and elsewhere in `txtail.go`?  If the latest round is `latest`, then transactions in block `latest` can have a `LastValid` of at most `latest+MaxTxnLife`.

Reply:
Once we've confirmed the `latest` round, we will start calling `isDup` on transactions whose `FirstValid` can be the next round, `latest+1`, so these transactions can have a `LastValid` of `latest+1+MaxTxnLife`. This initialization is done to return the correct value in `isDup` (and not return an error due to a missing `LastValid` in the map).

I guess initializing the map isn't strictly necessary here, since Go maps return `false` for membership checks if the map is `nil`. If I use Tsachi's suggestion below of keeping the last written round in the tracker, then I can probably remove this extra initialization.

Reply:
I think that the map initialization here is a good move. It ensures that on `newBlock` you will have exactly one new entry. ( it's relatively fast, predictable and have constant time ).
Note that loading from disk happens only on initialization, so performance in this function isn't an issue.

Reply:
>  so these transactions can have a LastValid of latest+1+MaxTxnLife.

goal did that when lastValid == 0. If firstVaild == 0 then we had validity window [lastRound+1, lastRound+1 + MaxTxnLife] that is MaxTxnLife + 1. I fixed that in goal.